### PR TITLE
Torch support for matching engines circus and OMP

### PR DESCRIPTION
**Pull request by yger**
This PR adds the possibility to use torch both for wobble and circus-omp matching engines. Torch is only used for the convolution, and even with a cpu backend, one can expect a x2 speedup compared to numpy/scipy. With cuda backend, the speedup is even better, but the problem is that one can not launch too many cores and/or should use the spawn method to create the processes

### CODE REVIEW

**Comment by yger** [comment]
Note that the PR depends on #3250

File: src/spikeinterface/sortingcomponents/clustering/split.py [hunk 1]
```diff
@@ -69,7 +69,7 @@ def split_clusters(
 
     original_labels = peak_labels
     peak_labels = peak_labels.copy()
-    split_count = np.zeros(peak_labels.size, dtype=int)
+    split_count = np.ones(peak_labels.size, dtype=int)
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/clustering/split.py:72]: ?????

File: src/spikeinterface/sortingcomponents/clustering/split.py [hunk 2]
```diff
@@ -187,7 +186,7 @@ def split(
         min_size_split=25,
         n_pca_features=2,
         scale_n_pca_by_depth=False,
-        minimum_common_channels=2,
+        sparsity_overlap_ratio=0.25,
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/clustering/split.py:189]:
```suggestion
        minimum_overlap_ratio=0.25,
```

File: src/spikeinterface/sortingcomponents/matching/circus.py [hunk 3]
```diff
@@ -140,13 +154,15 @@ def __init__(
...
-        max_failures=10,
+        max_failures=5,
         omp_min_sps=0.1,
         relative_error=5e-5,
         rank=5,
         ignore_inds=[],
-        vicinity=3,
+        vicinity=2,
         precomputed=None,
+        engine="numpy",
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/matching/circus.py:164]:
```suggestion
        engine="auto",
```

File: src/spikeinterface/sortingcomponents/matching/wobble.py [hunk 4]
```diff
@@ -41,6 +50,10 @@ class WobbleParameters:
         Maximum value for ampltiude scaling of templates.
     scale_amplitudes : bool
         If True, scale amplitudes of templates to match spikes.
+    engine : string in ["numpy", "torch", "auto"]. Default "auto"
+        The engine to use for the convolutions
+    torch_device : string in ["cpu", "cuda", "auto"]. Default "cpu"
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/matching/wobble.py:55]:
```suggestion
    torch_device : string in ["cpu", "cuda", None]. Default "cpu"
```

File: src/spikeinterface/sortingcomponents/peak_detection.py [hunk 5]
```diff
@@ -93,7 +123,13 @@ def detect_peaks(
     method_class = detect_peak_methods[method]
 
     method_kwargs, job_kwargs = split_job_kwargs(kwargs)
-    job_kwargs["mp_context"] = method_class.preferred_mp_context
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/peak_detection.py:96]: keep this. this was important for the detector based on torch

File: src/spikeinterface/sortingcomponents/peak_detection.py [hunk 6]
```diff
@@ -93,7 +123,13 @@ def detect_peaks(
...
+    _, torch_kwargs = split_job_kwargs(method_kwargs)
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/peak_detection.py:126]: why torch_kwarg would be slipted by split_job_kwargs ?

File: src/spikeinterface/benchmark/benchmark_matching.py [hunk 7]
```diff
@@ -74,6 +74,7 @@ def plot_performances_vs_snr(self, **kwargs):
     def plot_collisions(self, case_keys=None, figsize=None):
         if case_keys is None:
             case_keys = list(self.cases.keys())
+        import pylab as plt
```
Comment (samuelgarcia) [inline src/spikeinterface/benchmark/benchmark_matching.py:77]:
```suggestion
        import matplotlib.pyplot as plt
```

File: src/spikeinterface/sortingcomponents/matching/circus.py [hunk 8]
```diff
@@ -218,7 +247,7 @@ def _prepare_templates(self):
         self.max_similarity = np.zeros((self.num_templates, self.num_templates), dtype=np.float32)
         for i in range(self.num_templates):
             num_overlaps = np.sum(self.units_overlaps[i])
-            overlapping_units = np.flatnonzero(self.units_overlaps[i])
+            overlapping_units = np.where(self.units_overlaps[i])[0]
```
Comment (samuelgarcia) [inline src/spikeinterface/sortingcomponents/matching/circus.py:250]:
```suggestion
            overlapping_units = np.flatnonzero(self.units_overlaps[i])
```

**Review by samuelgarcia — APPROVED** [review]

**Pull request merged**